provider/oauth2.go

```diff
@@ -235,9 +247,108 @@ func (p Oauth2Handler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
 	p.JwtService.Reset(w)
 }
 
+func (p Oauth2Handler) loadUserFromIDToken(tok *oauth2.Token) (UserData, []byte, error) {
+	idToken, ok := tok.Extra("id_token").(string)
+	if !ok || idToken == "" {
+		return nil, nil, fmt.Errorf("id_token not found")
+	}
+
+	if p.keyfunc == nil {
+		err := p.tryInitJWKSKeyfunc()
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "can't load JWKS keys")
+		}
+	}
+
+	claims := jwt.MapClaims{}
+	parser := jwt.Parser{
+		// claims validation is not considering clock skew and randomly failing with iat validation
+		// nbf and exp are validated below
+		SkipClaimsValidation: true,
+	}
+
+	parsedIDToken, err := parser.ParseWithClaims(idToken, &claims, p.keyfunc)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to parse id token")
+	}
+
+	if !parsedIDToken.Valid {
+		return nil, nil, fmt.Errorf("invalid id token")
+	}
+
+	now := time.Now().Add(clockSkew).Unix()
+	if !claims.VerifyExpiresAt(now, false) {
+		return nil, nil, fmt.Errorf("id token expired")
+	}
+
+	if !claims.VerifyNotBefore(now, false) {
+		return nil, nil, fmt.Errorf("id token is not yet valid")
+	}
+
+	return UserData(claims), []byte(idToken), nil
+}
+
+func (p Oauth2Handler) loadUserFromEndpoint(client *http.Client) (UserData, []byte, error) {
+	uinfo, err := client.Get(p.infoURL)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to get client info")
+	}
+
+	defer func() {
+		if e := uinfo.Body.Close(); e != nil {
+			p.Logf("[WARN] failed to close response body, %s", e)
+		}
+	}()
+
+	data, err := io.ReadAll(uinfo.Body)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to read user info")
+	}
+
+	jData := map[string]interface{}{}
+	if e := json.Unmarshal(data, &jData); e != nil {
+		return nil, nil, errors.Wrap(e, "failed to unmarshal user info")
+	}
+	p.Logf("[DEBUG] got raw user info %+v", jData)
+
+	return jData, data, nil
+}
+
 func (p Oauth2Handler) makeRedirURL(path string) string {
 	elems := strings.Split(path, "/")
 	newPath := strings.Join(elems[:len(elems)-1], "/")
 
 	return strings.TrimSuffix(p.URL, "/") + strings.TrimSuffix(newPath, "/") + urlCallbackSuffix
 }
+
+func (p *Oauth2Handler) tryInitJWKSKeyfunc() error {
```

Comment:
I'm not sure what exactly this thing does, but it fills like the goal of this "try" is to set/init p.keyfunc once? If so, how about sync.Once? Will it make the intent more clear?